core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/common/AbstractUserStoreManager.java

```diff
@@ -11702,21 +11709,34 @@ protected boolean isUserIdGeneratedByUserStore(String userName, Map<String, Stri
      */
     public String getUserIDFromUserName(String userName) throws UserStoreException {
 
+        String userID;
         UserStore userStore = getUserStore(userName);
         if (userStore.isRecurssive()) {
             return ((AbstractUserStoreManager) userStore.getUserStoreManager())
                     .getUserIDFromUserName(userStore.getDomainFreeName());
         }
 
         if (isUniqueUserIdEnabledInUserStore(userStore)) {
-            return doGetUserIDFromUserNameWithID(userStore.getDomainFreeName());
+            userID = getFromUserIDCache(userName, userStore);
+            if (StringUtils.isEmpty(userID)) {
+                userID = doGetUserIDFromUserNameWithID(userName);
+                addToUserIDCache(userID, userName, userStore);
+            }
+            addToUserNameCache(userID, userName, userStore);
+            return userID;
         }
 
         Map<String, String> claims = doGetUserClaimValues(userName,
                 new String[]{UserCoreClaimConstants.USER_ID_CLAIM_URI},
                 userStore.getDomainName(), null);
         if (claims != null && claims.size() == 1) {
-            return claims.get(UserCoreClaimConstants.USER_ID_CLAIM_URI);
+            userID = getFromUserIDCache(userName, userStore);
+            if (StringUtils.isEmpty(userID)) {
+                userID = claims.get(UserCoreClaimConstants.USER_ID_CLAIM_URI);
```

Comment:
Better to move doGetUserClaimValues only if the cache is empty. Otherwise, that execution is redundant.

Reply:
Fixed by getting from the cache at the start